### DISCUSSION TITLE
How to develop/debug

### DISCUSSION THREAD

**Discussion by dannycarrera** [nginx-proxy/nginx-proxy#2415] [Q&A]
Hello, 

I was working on #1369 but not sure how to set breakpoints or log information specifically from the nginx.tmpl file.

**Comment by buchdag** [comment]
Hi @dannycarrera 

I must admit that the template is getting harder and harder to work on if you're not used to it, and the fact that we have almost zero language feature with Go template (at least with VS Code) is not making it easier.

I personally use [this VS Code plugin](https://marketplace.visualstudio.com/items?itemName=casualjim.gotemplate) and set `nginx.tmpl` to **Golang Text Template (gotemplate)**. This isn't the same as **Go Template File (gotmpl)** provided by the official Go Extension, which tend to be selected by default and in this case does not work at all.

![image](https://github.com/nginx-proxy/nginx-proxy/assets/13685826/e65bb3b3-050f-4a05-93a7-af28f1814d2b)

I also usually mount the template file from source to the running `nginx-proxy` container rather than rebuilding the image every time (with `-v /path/to/nginx.tmpl:/app/nginx.tmpl:ro` or its compose equivalent).

If you need to log information you'll have to render them as comment on the template, for instance logging `$globals` to the rendered template:

```
# {{ printf "%#v" $globals }}
```

You can also render a variable content as JSON, but be advised that will result in a faulty nginx conf with variables having a too large content (ie `$globals`):

```
# {{ toJson $foo }}
```

Since this is a templating language there isn't a notion of breakpoint, but you can use Sprig's [`fail`](https://masterminds.github.io/sprig/flow_control.html) function to conditionally halt execution. This example will halt execution if `someCondition` is `true`, and print the entire top level context as pretty JSON to the container logs:

```
{{- if someCondition }}
  {{- fail (toPrettyJson $) }}
{{- end }}
```

To print the current scoped context, use `.` instead of `$`:
```
{{- if someCondition }}
  {{- fail (toPrettyJson .) }}
{{- end }}
```

Maybe we could add a docker-gen function to help render commented out pretty JSON ? 🤔

**Reply by buchdag** [reply]
> Maybe we could add a docker-gen function to help render commented out pretty JSON

Done in https://github.com/nginx-proxy/docker-gen/pull/612, I'll notify here when this is made available in nginx-proxy.
